Fix 367

dingquanyu opened the pull request:
1. moved the step of dumping pickles into postprocessing after all the metrics are calculated
2. added test on the contents inside the result pickle
3. fixed #367

dingquanyu commented:
Unittests on testRun1 and testRun2 were successful. I believe it's safe to merge.

dingquanyu commented on test/check_predict_structure.py:
```diff
@@ -76,6 +77,14 @@ def _runCommonTests(self, result):
+        if multimer_mode:
+            required_keys = ['distogram', 'experimentally_resolved', 'masked_msa', 'predicted_aligned_error', 'predicted_lddt', 'structure_module', 'plddt', 'aligned_confidence_probs', 'max_predicted_aligned_error', 'seqs', 'iptm', 'ranking_confidence']
+        else:
+            required_keys = ['distogram', 'experimentally_resolved', 'masked_msa', 'predicted_aligned_error', 'predicted_lddt', 'structure_module', 'plddt', 'aligned_confidence_probs', 'max_predicted_aligned_error', 'seqs','ranking_confidence']
```
At the beginning I thought ptm will be one of these but it turned out ptm never 

> you don't check for 'ptm' here?

oh that's true. ptm is reported by AlphaFold as well. Forgot to add this

DimaMolod commented:
> 1. moved the step of dumping pickles into postprocessing after all the metrics are calculated
> 2. added test on the contents inside the result pickle
> 3. fixed ['ranking_confidence' is not in the result.pkl file after the update #367](https://github.com/KosinskiLab/AlphaPulldown/issues/367)

if you dump pickles after all predictions are made, the resume (for prediction) will not work because it reads the prediction results from pkls here
https://github.com/KosinskiLab/AlphaPulldown/blob/fix-367/alphapulldown/folding_backend/alphafold_backend.py#L302

dingquanyu commented:
> > 1. moved the step of dumping pickles into postprocessing after all the metrics are calculated
> > 2. added test on the contents inside the result pickle
> > 3. fixed ['ranking_confidence' is not in the result.pkl file after the update #367](https://github.com/KosinskiLab/AlphaPulldown/issues/367)
> 
> if you dump pickles after all predictions are made, the resume (for prediction) will not work because it reads the prediction results from pkls here https://github.com/KosinskiLab/AlphaPulldown/blob/fix-367/alphapulldown/folding_backend/alphafold_backend.py#L302

That's true. Maybe dump pickle as well in the prediction step? Anyway the pickles are gonna be overwritten again in the postprocess step.

DimaMolod commented:
yes, dumping pickles twice would be the easiest solution, I guess. Could you please make sure that testRun_5 works fine after the update?

dingquanyu commented:
> yes, dumping pickles twice would be the easiest solution, I guess. Could you please make sure that testRun_5 works fine after the update?

OK! Tests are running now

dingquanyu commented:
> yes, dumping pickles twice would be the easiest solution, I guess. Could you please make sure that testRun_5 works fine after the update?

Now test 1,2 and 5 are all successful. That was a good catch btw.  I think now it's safe to merge
